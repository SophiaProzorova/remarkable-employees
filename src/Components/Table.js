import { useMemo } from "react";
import { useSortBy, useTable } from "react-table";
import { COLUMNS } from "./columns"

export default function Table (props){
    // import columns title from file
    const columnsTitle = useMemo(()=> COLUMNS, []);

    // create cookies for safe sorting data
    
    let desc = (localStorage.getItem("column") == 0) ? false : true;
    

    const initialState = {
        sortBy:[
            {
                id: localStorage.getItem("column"),
                desc: desc
            }
        ]
    }

    const tableInstance = useTable({
        columns: columnsTitle,
        data: props.values,
        initialState
    }, useSortBy)

    const { 
        getTableProps, 
        getTableBodyProps, 
        headerGroups, 
        rows, 
        prepareRow
    } = tableInstance;

    return (
        <table  {...getTableProps()} className="tableInfo">
            <thead>
                {
                    headerGroups.map(headerGroup => (
                        <tr {...headerGroup.getHeaderGroupProps()}>
                            {
                                headerGroup.headers.map((column, index) => {
                                    let id = "", desc="";
                                    if (column.isSorted) {
                                        id = column.id;
                                        desc = column.isSortedDesc ? '1' : '0'
                                    }
                                    if (id!=="") {
                                        localStorage.setItem("column", id);
                                        localStorage.setItem("desc", desc)
                                    }
                                    return (<th {...column.getHeaderProps(column.getSortByToggleProps())} >
                                        {
                                            column.render('Header')
                                        }
                                    </th>)
                                })
                            }
                        </tr>
                    ))
                }
            </thead>
            <tbody {...getTableBodyProps()}>
                {
                    rows.map( row => {
                        prepareRow(row)
                        return (
                            <tr {...row.getRowProps()}>
                                {   
                                    
                                    row.cells.map( cell => {
                                        return <td {...cell.getCellProps({className:cell.column.className})}>{cell.render('Cell')}</td>
                                    })
                                }
                            </tr>
                        )
                    })
                }
            </tbody>
        </table>    
    );
}