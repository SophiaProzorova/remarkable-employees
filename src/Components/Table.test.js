import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event';
import Table from './Table'

describe('renders a table', ()=>{

    it('render a table with props', ()=>{
        const data =[
            {
                "name": "Jesse Karl",
                "jobTitle": "Developer",
                "tenure": "3",
                "gender": "Male"
              },
              {
                "name": "Kelly Turner",
                "jobTitle": "QA",
                "tenure": "8",
                "gender": "Female"
              }
        ]

        render(< Table values = {data} />);
        expect(screen.getAllByRole('row')[1]).toHaveTextContent("Kelly Turner");
    });

    it('sort column on click', ()=>{
        const data =[
            {
                "name": "Kelly Turner",
                "jobTitle": "QA",
                "tenure": "8",
                "gender": "Female"
            },
            {
                "name": "Jesse Karl",
                "jobTitle": "Developer",
                "tenure": "3",
                "gender": "Male"
              }
        ]

        render(< Table values = {data} />);
        expect(screen.getAllByRole('row')[2]).toHaveTextContent("QA");
        const column = screen.getByText(/job title/i);
        userEvent.click(column);
        expect(screen.getAllByRole('row')[1]).toHaveTextContent("Developer");
    });

    it('Table snapshot', ()=>{
        const data =[
            {
                "name": "Kelly Turner",
                "jobTitle": "QA",
                "tenure": "8",
                "gender": "Female"
            },
            {
                "name": "Jesse Karl",
                "jobTitle": "Developer",
                "tenure": "3",
                "gender": "Male"
              }
        ]
        const table = render (<Table values={data} />)
        expect(table).toMatchSnapshot();
    })
})