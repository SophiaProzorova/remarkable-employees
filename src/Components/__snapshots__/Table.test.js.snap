// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a table Table snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <table
        class="tableInfo"
        role="table"
      >
        <thead>
          <tr
            role="row"
          >
            <th
              colspan="1"
              role="columnheader"
              style="cursor: pointer;"
              title="Toggle SortBy"
            >
              Name
            </th>
            <th
              colspan="1"
              role="columnheader"
              style="cursor: pointer;"
              title="Toggle SortBy"
            >
              Job title
            </th>
            <th
              colspan="1"
              role="columnheader"
              style="cursor: pointer;"
              title="Toggle SortBy"
            >
              Tenure
            </th>
            <th
              colspan="1"
              role="columnheader"
              style="cursor: pointer;"
              title="Toggle SortBy"
            >
              Gender
            </th>
          </tr>
        </thead>
        <tbody
          role="rowgroup"
        >
          <tr
            role="row"
          >
            <td
              role="cell"
            >
              Kelly Turner
            </td>
            <td
              role="cell"
            >
              QA
            </td>
            <td
              role="cell"
            >
              8
            </td>
            <td
              role="cell"
            >
              Female
            </td>
          </tr>
          <tr
            role="row"
          >
            <td
              role="cell"
            >
              Jesse Karl
            </td>
            <td
              role="cell"
            >
              Developer
            </td>
            <td
              role="cell"
            >
              3
            </td>
            <td
              role="cell"
            >
              Male
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </body>,
  "container": <div>
    <table
      class="tableInfo"
      role="table"
    >
      <thead>
        <tr
          role="row"
        >
          <th
            colspan="1"
            role="columnheader"
            style="cursor: pointer;"
            title="Toggle SortBy"
          >
            Name
          </th>
          <th
            colspan="1"
            role="columnheader"
            style="cursor: pointer;"
            title="Toggle SortBy"
          >
            Job title
          </th>
          <th
            colspan="1"
            role="columnheader"
            style="cursor: pointer;"
            title="Toggle SortBy"
          >
            Tenure
          </th>
          <th
            colspan="1"
            role="columnheader"
            style="cursor: pointer;"
            title="Toggle SortBy"
          >
            Gender
          </th>
        </tr>
      </thead>
      <tbody
        role="rowgroup"
      >
        <tr
          role="row"
        >
          <td
            role="cell"
          >
            Kelly Turner
          </td>
          <td
            role="cell"
          >
            QA
          </td>
          <td
            role="cell"
          >
            8
          </td>
          <td
            role="cell"
          >
            Female
          </td>
        </tr>
        <tr
          role="row"
        >
          <td
            role="cell"
          >
            Jesse Karl
          </td>
          <td
            role="cell"
          >
            Developer
          </td>
          <td
            role="cell"
          >
            3
          </td>
          <td
            role="cell"
          >
            Male
          </td>
        </tr>
      </tbody>
    </table>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
